File: ppl-meta-communications/src/routes/email_settings.py
"""
Email Settings API routes for managing SMTP configuration.
"""
import logging
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from ..database import get_db
from ..models.email_settings import EmailSettings
from ..schemas.email_settings import (
    EmailSettingsCreate,
    EmailSettingsUpdate,
    EmailSettingsResponse,
)

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/v1/settings/email", tags=["email-settings"])


@router.get("", response_model=EmailSettingsResponse)
async def get_email_settings(db: Session = Depends(get_db)):
    """
    Get current email settings.
    
    Returns the first (and only) email settings record.
    Creates default settings if none exist.
    """
    settings = db.query(EmailSettings).first()
    
    if not settings:
        # Create default settings
        settings = EmailSettings()
        db.add(settings)
        db.commit()
        db.refresh(settings)
        logger.info("Created default email settings")
    
    # Mask password in response
    response_dict = {
        "id": settings.id,
        "mail_enabled": settings.mail_enabled,
        "mail_server": settings.mail_server,
        "mail_port": settings.mail_port,
        "mail_username": settings.mail_username,
        "mail_password": "********" if settings.mail_password else "",
        "mail_from": settings.mail_from,
        "mail_from_name": settings.mail_from_name,
        "mail_starttls": settings.mail_starttls,
        "mail_ssl_tls": settings.mail_ssl_tls,
        "use_credentials": settings.use_credentials,
        "created_at": settings.created_at,
        "updated_at": settings.updated_at,
    }
    
    return EmailSettingsResponse(**response_dict)


@router.put("", response_model=EmailSettingsResponse)
async def update_email_settings(
    settings_update: EmailSettingsUpdate,
    db: Session = Depends(get_db)
):
    """
    Update email settings.
    
    Updates the email SMTP configuration. Only provided fields will be updated.
    """
    settings = db.query(EmailSettings).first()
    
    if not settings:
        # Create new settings if none exist
        settings = EmailSettings()
        db.add(settings)
    
    # Update only provided fields
    update_data = settings_update.model_dump(exclude_unset=True)
    
    for field, value in update_data.items():
        if value is not None:  # Only update if value is explicitly provided
            setattr(settings, field, value)
    
    db.commit()
    db.refresh(settings)
    
    logger.info(f"Updated email settings: mail_enabled={settings.mail_enabled}, server={settings.mail_server}")
    
    # Mask password in response
    response_dict = {
        "id": settings.id,
        "mail_enabled": settings.mail_enabled,
        "mail_server": settings.mail_server,
        "mail_port": settings.mail_port,
        "mail_username": settings.mail_username,
        "mail_password": "********" if settings.mail_password else "",
        "mail_from": settings.mail_from,
        "mail_from_name": settings.mail_from_name,
        "mail_starttls": settings.mail_starttls,
        "mail_ssl_tls": settings.mail_ssl_tls,
        "use_credentials": settings.use_credentials,
        "created_at": settings.created_at,
        "updated_at": settings.updated_at,
    }
    
    return EmailSettingsResponse(**response_dict)


@router.post("/test", response_model=dict)
async def test_email_settings(
    test_email: str,
    db: Session = Depends(get_db)
):
    """
    Test email settings by sending a test email.
    
    Args:
        test_email: Email address to send test email to
        
    Returns:
        Success status and message
    """
    settings = db.query(EmailSettings).first()
    
    if not settings or not settings.mail_enabled:
        raise HTTPException(
            status_code=400,
            detail="Email is not enabled. Please enable and configure email settings first."
        )
    
    # Check if all required fields are configured
    if not settings.mail_server or not settings.mail_username or not settings.mail_from:
        raise HTTPException(
            status_code=400,
            detail="Email settings are incomplete. Please configure SMTP server, username, and from address."
        )
    
    # Import email service
    from ..services.email_service import EmailService
    
    try:
        email_service = EmailService(db)
        
        # Send test email
        success, message, log_uuid = await email_service.send_email(
            to=[test_email],
            subject="PPL Meta Email Test",
            text_body="This is a test email from PPL Meta Communications Service. If you received this, your email settings are configured correctly!",
        )
        
        if success:
            return {
                "success": True,
                "message": f"Test email sent successfully to {test_email}"
            }
        else:
            raise HTTPException(
                status_code=500,
                detail=f"Failed to send test email: {message}"
            )
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error testing email settings: {e}", exc_info=True)
        raise HTTPException(
            status_code=500,
            detail=f"Error sending test email: {str(e)}"
        )

File: ppl-meta-frontend/lib/models/email_settings.dart
// Email settings model for SMTP configuration
import 'package:json_annotation/json_annotation.dart';

part 'email_settings.g.dart';

@JsonSerializable()
class EmailSettings {
  final int id;
  @JsonKey(name: 'mail_enabled')
  final bool mailEnabled;
  @JsonKey(name: 'mail_server')
  final String mailServer;
  @JsonKey(name: 'mail_port')
  final int mailPort;
  @JsonKey(name: 'mail_username')
  final String mailUsername;
  @JsonKey(name: 'mail_password')
  final String mailPassword;
  @JsonKey(name: 'mail_from')
  final String mailFrom;
  @JsonKey(name: 'mail_from_name')
  final String mailFromName;
  @JsonKey(name: 'mail_starttls')
  final bool mailStarttls;
  @JsonKey(name: 'mail_ssl_tls')
  final bool mailSslTls;
  @JsonKey(name: 'use_credentials')
  final bool useCredentials;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;

  const EmailSettings({
    required this.id,
    required this.mailEnabled,
    required this.mailServer,
    required this.mailPort,
    required this.mailUsername,
    required this.mailPassword,
    required this.mailFrom,
    required this.mailFromName,
    required this.mailStarttls,
    required this.mailSslTls,
    required this.useCredentials,
    required this.createdAt,
    required this.updatedAt,
  });

  factory EmailSettings.fromJson(Map<String, dynamic> json) =>
      _$EmailSettingsFromJson(json);

  Map<String, dynamic> toJson() => _$EmailSettingsToJson(this);

  EmailSettings copyWith({
    int? id,
    bool? mailEnabled,
    String? mailServer,
    int? mailPort,
    String? mailUsername,
    String? mailPassword,
    String? mailFrom,
    String? mailFromName,
    bool? mailStarttls,
    bool? mailSslTls,
    bool? useCredentials,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return EmailSettings(
      id: id ?? this.id,
      mailEnabled: mailEnabled ?? this.mailEnabled,
      mailServer: mailServer ?? this.mailServer,
      mailPort: mailPort ?? this.mailPort,
      mailUsername: mailUsername ?? this.mailUsername,
      mailPassword: mailPassword ?? this.mailPassword,
      mailFrom: mailFrom ?? this.mailFrom,
      mailFromName: mailFromName ?? this.mailFromName,
      mailStarttls: mailStarttls ?? this.mailStarttls,
      mailSslTls: mailSslTls ?? this.mailSslTls,
      useCredentials: useCredentials ?? this.useCredentials,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

@JsonSerializable()
class EmailSettingsUpdate {
  @JsonKey(name: 'mail_enabled')
  final bool? mailEnabled;
  @JsonKey(name: 'mail_server')
  final String? mailServer;
  @JsonKey(name: 'mail_port')
  final int? mailPort;
  @JsonKey(name: 'mail_username')
  final String? mailUsername;
  @JsonKey(name: 'mail_password')
  final String? mailPassword;
  @JsonKey(name: 'mail_from')
  final String? mailFrom;
  @JsonKey(name: 'mail_from_name')
  final String? mailFromName;
  @JsonKey(name: 'mail_starttls')
  final bool? mailStarttls;
  @JsonKey(name: 'mail_ssl_tls')
  final bool? mailSslTls;
  @JsonKey(name: 'use_credentials')
  final bool? useCredentials;

  const EmailSettingsUpdate({
    this.mailEnabled,
    this.mailServer,
    this.mailPort,
    this.mailUsername,
    this.mailPassword,
    this.mailFrom,
    this.mailFromName,
    this.mailStarttls,
    this.mailSslTls,
    this.useCredentials,
  });

  factory EmailSettingsUpdate.fromJson(Map<String, dynamic> json) =>
      _$EmailSettingsUpdateFromJson(json);

  Map<String, dynamic> toJson() => _$EmailSettingsUpdateToJson(this);
}
